vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+	} else {
+		i.oidcCache.lock.RUnlock()
+
+		c, err := i.getOIDCConfig(ctx, req.Storage)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: review required contents
+		disc := discovery{
+			Issuer:      c.effectiveIssuer,
+			Keys:        c.effectiveIssuer + "/.well-known/keys",
+			Subjects:    []string{"public"},
+			IDTokenAlgs: []string{string(jose.RS256)},
+		}
+
+		data, err = json.Marshal(disc)
+		if err != nil {
+			return nil, err
+		}
+
+		i.oidcCache.lock.Lock()
+		i.oidcCache.discoveryResponse = data
+		i.oidcCache.lock.Unlock()
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			logical.HTTPStatusCode:  200,
+			logical.HTTPRawBody:     data,
+			logical.HTTPContentType: "application/json",
+		},
+	}
+
+	return resp, nil
+}
+
+// pathOIDCReadPublicKeys is used to retrieve all public keys so that clients can
+// verify the validity of a signed OIDC token.
+func (i *IdentityStore) pathOIDCReadPublicKeys(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var data []byte
+
+	i.oidcCache.lock.RLock()
+	if len(i.oidcCache.jwksResponse) > 0 {
+		data = i.oidcCache.jwksResponse
+		i.oidcCache.lock.RUnlock()
+	} else {
+		i.oidcCache.lock.RUnlock()
+
+		keyIDs, err := listOIDCPublicKeys(ctx, req.Storage)
+		if err != nil {
+			return nil, err
+		}
+
+		jwks := jose.JSONWebKeySet{
+			Keys: make([]jose.JSONWebKey, 0, len(keyIDs)),
+		}
+
+		for _, keyID := range keyIDs {
+			key, err := loadOIDCPublicKey(ctx, req.Storage, keyID)
+			if err != nil {
+				return nil, err
+			}
+			jwks.Keys = append(jwks.Keys, *key)
+		}
+
+		data, err = json.Marshal(jwks)
+		if err != nil {
+			return nil, err
+		}
+
+		i.oidcCache.lock.Lock()
+		i.oidcCache.jwksResponse = data
+		i.oidcCache.lock.Unlock()
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			logical.HTTPStatusCode:  200,
+			logical.HTTPRawBody:     data,
+			logical.HTTPContentType: "application/json",
+		},
+	}
+
+	return resp, nil
+}
+
+func (i *IdentityStore) pathOIDCIntrospect(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var claims jwt.Claims
+	var errResponse string
+
+	rawIDToken := d.Get("token").(string)
+	clientID := d.Get("client_id").(string)
+
+	// validate basic JWT structure
+	parsedJWT, err := jwt.ParseSigned(rawIDToken)
+	if err != nil {
+		errResponse = fmt.Sprintf("error parsing token: %s", err.Error())
+	}
+
+	// validate signature
+	if errResponse == "" {
+		keyIDs, err := listOIDCPublicKeys(ctx, req.Storage) // TODO: use cached keys
+		if err != nil {
+			return nil, err
+		}
+
+		var valid bool
+		for _, keyID := range keyIDs {
+			key, err := loadOIDCPublicKey(ctx, req.Storage, keyID)
+			if err != nil {
+				return nil, err
+			}
+			if err := parsedJWT.Claims(key, &claims); err == nil {
+				valid = true
+				break
+			}
+		}
+
+		if !valid {
+			errResponse = "unable to validate the token signature"
```

Comment:
There are a bunch of `if errResponse == ""` statements below, is it not feasible to return earlier?

Reply:
This was done mainly because the response is non-standard (i.e. many lines to prepare). But yeah it's ugly. I've now factored it into a nested function for preparing the response, so we can return at the point of the error. I think it's a little cleaner.

Reply:
e.g.
```
	// helper for preparing the non-standard introspection response
	introspectionResp := func(errorMsg string) (*logical.Response, error) {
		response := map[string]interface{}{
			"active": true,
		}

		if errorMsg != "" {
			response["active"] = false
			response["error"] = errorMsg
		}

		data, err := json.Marshal(response)
		if err != nil {
			return nil, err
		}

		resp := &logical.Response{
			Data: map[string]interface{}{
				logical.HTTPStatusCode:  200,
				logical.HTTPRawBody:     data,
				logical.HTTPContentType: "application/json",
			},
		}

		return resp, nil
	}
```

Reply:
Updated